### PR TITLE
Add more detailed shard status on shutdown

### PR DESCRIPTION
**Pull request by rjernst**
When the shutdown api is polled, shard migrations are reported as a single total remaining. However, that number is made up of more detailed counts within cluster state. This commit expands the counts reported for migration status to include started, relocating and initializing shard counts.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java [hunk 1]
```diff
@@ -36,35 +36,90 @@ public class ShutdownShardMigrationStatus implements Writeable, ChunkedToXConten
...
+            startedShards,
+            relocatingShards,
+            initializingShards,
+            startedShards + relocatingShards + initializingShards,
+            explanation,
+            null
+        );
+    }
+
+    public ShutdownShardMigrationStatus(
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java:92]: Can this constructor be private? The constructors above this one enforce the relationship between shardsRemaining and started/relocating/initializing, either defining shardsRemaining as the sum or setting the others to -1.

**Review by williamrandolph — CHANGES_REQUESTED** [review]
I just have one small change to request.

File: server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java (hunk 1 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java]: Seems like it should work. I pushed [fa15dd9](https://github.com/elastic/elasticsearch/pull/102271/commits/fa15dd9bc3eda5b8dbe2e4a1b9aece0b532abdae).

**Review by williamrandolph — APPROVED** [review]
LGTM!

File: server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java [hunk 2]
```diff
@@ -36,35 +36,90 @@ public class ShutdownShardMigrationStatus implements Writeable, ChunkedToXConten
...
     @Nullable
     private final ShardAllocationDecision allocationDecision;
 
-    public ShutdownShardMigrationStatus(SingleNodeShutdownMetadata.Status status, long shardsRemaining) {
-        this(status, shardsRemaining, null, null);
+    public ShutdownShardMigrationStatus(
+        SingleNodeShutdownMetadata.Status status,
+        long shardsRemaining,
+        @Nullable String explanation,
+        @Nullable ShardAllocationDecision allocationDecision
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java:52]: Is this parameter intentionally not used in constructor body?

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java:52]: This was unintentional. I opened https://github.com/elastic/elasticsearch/pull/105399 to fix. Thanks for noticing.

**Pull request merged**